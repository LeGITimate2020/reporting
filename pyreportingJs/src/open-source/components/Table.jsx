import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import * as d3 from 'd3';
import ReactTable from 'react-table';

import { calculateDomain, textColor } from '../utils/heatmapHelpers';

const MAX_DEFAULT_PAGE_SIZE = 10;
const DEFAULT_COLOR_RANGE = ['#ff9980', '#ffffff', '#8099ff'];


/**
 * Extends ReactTable functionality with conditional formatting
 */
class Table extends React.Component {
  /**
   * returns a d3 color scale for individual scalar values in the table
   * @return {Object} - { all: d3ColorScale, columns: { 'col A': d3ColorScale, ...} }
   *
   */
  colorScalesFromData = () => {
    const { data } = this.props;
    const colorScales = { all: null, columns: {} };

    // scan over all values and columns, bucketing into all and by column
    const allValues = [];
    const allColumns = Object.keys(data[0]);
    const columnToValues = {};
    for (const column of allColumns) {
      for (const row of data) {
        const value = _.get(row, column);
        if (!isNaN(Number(value))) {
          if (this.props.colorByColumn) {
            if (!columnToValues[column]) {
              columnToValues[column] = [];
            }
            columnToValues[column].push({ value });
          }
          allValues.push({ value });
        }
      }
    }

    colorScales.all = d3.scaleLinear()
      .domain(calculateDomain(allValues))
      .range(this.props.colorRange)
      .clamp(true);

    if (this.props.colorByColumn) {
      for (const [column, values] of Object.entries(columnToValues)) {
        // TODO: generate N different color gradients for
        // N columns. v0: enumerate primary color pairs. v1: generate N color gradients
        // with largest possible distance between them in HSV space.
        // perhaps w/ http://gka.github.io/chroma.js/
        colorScales.columns[column] = d3.scaleLinear()
          .domain(calculateDomain(values))
          .range(this.props.colorRange)
          .clamp(true);
      }
    }

    return colorScales;
  };

  getTdProps = (state, rowInfo, column) => {
    if (!rowInfo || !column) {
      return {};
    }
    if (!this.colorScales) {
      this.colorScales = this.colorScalesFromData();
    }
    const value = rowInfo.row[column.id];
    let colorScale = null;
    if (this.props.colorByColumn) {
      colorScale = (this.colorScales.columns[column.id] || (() => ('')));
    } else {
      colorScale = this.colorScales.all;
    }
    const backgroundColor = isNaN(Number(value)) ? '' : colorScale(value);
    const color = textColor(backgroundColor);
    return {
      style: { backgroundColor, color },
    };
  };

  render() {
    const {
      title, renderEmptyChart, data, columns, reactTableProps,
    } = this.props;

    const haveEnoughToRender = columns.length > 0;
    if (!haveEnoughToRender && !renderEmptyChart) {
      return <div />;
    }
    return (
      <div>
        <h4 style={{ textAlign: 'center' }}>
          {title}
        </h4>
        <ReactTable
          data={data}
          columns={columns}
          className="allow-wrap compact"
          defaultPageSize={10}
          getTdProps={this.getTdProps}
          showPaginationBottom={data.length > MAX_DEFAULT_PAGE_SIZE}
          {...reactTableProps}
        />
      </div>
    );
  }
}

Table.propTypes = {
  title: PropTypes.string.isRequired,
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
  columns: PropTypes.arrayOf(PropTypes.object).isRequired,
  colorRange: PropTypes.arrayOf(PropTypes.string),
  colorByColumn: PropTypes.bool,
  reactTableProps: PropTypes.object,
  renderEmptyChart: PropTypes.bool,
};

Table.defaultProps = {
  colorRange: DEFAULT_COLOR_RANGE,
  colorByColumn: false,
  reactTableProps: {},
  renderEmptyChart: false,
};

export default Table;
