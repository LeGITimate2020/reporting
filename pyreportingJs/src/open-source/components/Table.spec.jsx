import React from 'react';
import { shallow } from 'enzyme';
import Table from './Table';

describe('Table.jsx', () => {
  const data = [
    {
      column1: 11,
      column2: 12,
      column3: 'abc',
    },
    {
      column1: 21,
      column2: 22,
      column3: 'def',
    },
  ];
  const columns = ['column1', 'column2', 'column3'];

  const reactTableColumns = columns.map(column => ({ Header: column, accessor: column }));

  const props = {
    columns: reactTableColumns,
    data,
    title: 'testable test table',
  };

  it('renders without crashing!', () => {
    shallow(<Table {...props} />);
  });

  it('Contains one ReactTable and correct h4 title', () => {
    const wrapper = shallow(<Table {...props} />);
    expect(wrapper.find('h4').text()).toBe(props.title);
    expect(wrapper.find('ReactTable').length).toBe(1);
  });

  it('Matches snapshot', () => {
    const wrapper = shallow(<Table {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
