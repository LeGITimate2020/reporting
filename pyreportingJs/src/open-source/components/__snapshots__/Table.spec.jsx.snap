// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table.jsx Matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Table
    colorByColumn={false}
    colorRange={
      Array [
        "#ff9980",
        "#ffffff",
        "#8099ff",
      ]
    }
    columns={
      Array [
        Object {
          "Header": "column1",
          "accessor": "column1",
        },
        Object {
          "Header": "column2",
          "accessor": "column2",
        },
        Object {
          "Header": "column3",
          "accessor": "column3",
        },
      ]
    }
    data={
      Array [
        Object {
          "column1": 11,
          "column2": 12,
          "column3": "abc",
        },
        Object {
          "column1": 21,
          "column2": 22,
          "column3": "def",
        },
      ]
    }
    reactTableProps={Object {}}
    renderEmptyChart={false}
    title="testable test table"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h4
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          testable test table
        </h4>,
        <ReactTable
          AggregatedComponent={[Function]}
          ExpanderComponent={[Function]}
          FilterComponent={[Function]}
          LoadingComponent={[Function]}
          NextComponent={undefined}
          NoDataComponent={[Function]}
          PadRowComponent={[Function]}
          PaginationComponent={[Function]}
          PivotComponent={undefined}
          PivotValueComponent={[Function]}
          PreviousComponent={undefined}
          ResizerComponent={[Function]}
          TableComponent={[Function]}
          TbodyComponent={[Function]}
          TdComponent={[Function]}
          TfootComponent={[Function]}
          ThComponent={[Function]}
          TheadComponent={[Function]}
          TrComponent={[Function]}
          TrGroupComponent={[Function]}
          aggregatedKey="_aggregated"
          className="allow-wrap compact"
          collapseOnDataChange={true}
          collapseOnPageChange={true}
          collapseOnSortingChange={true}
          column={
            Object {
              "Aggregated": undefined,
              "Cell": undefined,
              "Expander": undefined,
              "Filter": undefined,
              "Footer": undefined,
              "Header": undefined,
              "Pivot": undefined,
              "PivotValue": undefined,
              "aggregate": undefined,
              "className": "",
              "filterAll": false,
              "filterMethod": undefined,
              "filterable": undefined,
              "footerClassName": "",
              "footerStyle": Object {},
              "getFooterProps": [Function],
              "getHeaderProps": [Function],
              "getProps": [Function],
              "headerClassName": "",
              "headerStyle": Object {},
              "minWidth": 100,
              "resizable": undefined,
              "show": true,
              "sortMethod": undefined,
              "sortable": undefined,
              "style": Object {},
            }
          }
          columns={
            Array [
              Object {
                "Header": "column1",
                "accessor": "column1",
              },
              Object {
                "Header": "column2",
                "accessor": "column2",
              },
              Object {
                "Header": "column3",
                "accessor": "column3",
              },
            ]
          }
          data={
            Array [
              Object {
                "column1": 11,
                "column2": 12,
                "column3": "abc",
              },
              Object {
                "column1": 21,
                "column2": 22,
                "column3": "def",
              },
            ]
          }
          defaultExpanded={Object {}}
          defaultFilterMethod={[Function]}
          defaultFiltered={Array []}
          defaultPageSize={10}
          defaultResized={Array []}
          defaultSortDesc={false}
          defaultSortMethod={[Function]}
          defaultSorted={Array []}
          expanderDefaults={
            Object {
              "filterable": false,
              "resizable": false,
              "sortable": false,
              "width": 35,
            }
          }
          filterable={false}
          freezeWhenExpanded={false}
          getLoadingProps={[Function]}
          getNoDataProps={[Function]}
          getPaginationProps={[Function]}
          getProps={[Function]}
          getResizerProps={[Function]}
          getTableProps={[Function]}
          getTbodyProps={[Function]}
          getTdProps={[Function]}
          getTfootProps={[Function]}
          getTfootTdProps={[Function]}
          getTfootTrProps={[Function]}
          getTheadFilterProps={[Function]}
          getTheadFilterThProps={[Function]}
          getTheadFilterTrProps={[Function]}
          getTheadGroupProps={[Function]}
          getTheadGroupThProps={[Function]}
          getTheadGroupTrProps={[Function]}
          getTheadProps={[Function]}
          getTheadThProps={[Function]}
          getTheadTrProps={[Function]}
          getTrGroupProps={[Function]}
          getTrProps={[Function]}
          groupedByPivotKey="_groupedByPivot"
          indexKey="_index"
          loading={false}
          loadingText="Loading..."
          multiSort={true}
          nestingLevelKey="_nestingLevel"
          nextText="Next"
          noDataText="No rows found"
          ofText="of"
          onExpandedChange={undefined}
          onFetchData={[Function]}
          onFilteredChange={undefined}
          onPageChange={undefined}
          onPageSizeChange={undefined}
          onResizedChange={undefined}
          onSortedChange={undefined}
          originalKey="_original"
          pageSizeOptions={
            Array [
              5,
              10,
              20,
              25,
              50,
              100,
            ]
          }
          pageText="Page"
          pivotBy={undefined}
          pivotDefaults={Object {}}
          pivotIDKey="_pivotID"
          pivotValKey="_pivotVal"
          previousText="Previous"
          resizable={true}
          resolveData={[Function]}
          rowsText="rows"
          showPageJump={true}
          showPageSizeOptions={true}
          showPagination={true}
          showPaginationBottom={false}
          showPaginationTop={false}
          sortable={true}
          style={Object {}}
          subRowsKey="_subRows"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "testable test table",
          "style": Object {
            "textAlign": "center",
          },
        },
        "ref": null,
        "rendered": "testable test table",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "AggregatedComponent": [Function],
          "ExpanderComponent": [Function],
          "FilterComponent": [Function],
          "LoadingComponent": [Function],
          "NextComponent": undefined,
          "NoDataComponent": [Function],
          "PadRowComponent": [Function],
          "PaginationComponent": [Function],
          "PivotComponent": undefined,
          "PivotValueComponent": [Function],
          "PreviousComponent": undefined,
          "ResizerComponent": [Function],
          "TableComponent": [Function],
          "TbodyComponent": [Function],
          "TdComponent": [Function],
          "TfootComponent": [Function],
          "ThComponent": [Function],
          "TheadComponent": [Function],
          "TrComponent": [Function],
          "TrGroupComponent": [Function],
          "aggregatedKey": "_aggregated",
          "className": "allow-wrap compact",
          "collapseOnDataChange": true,
          "collapseOnPageChange": true,
          "collapseOnSortingChange": true,
          "column": Object {
            "Aggregated": undefined,
            "Cell": undefined,
            "Expander": undefined,
            "Filter": undefined,
            "Footer": undefined,
            "Header": undefined,
            "Pivot": undefined,
            "PivotValue": undefined,
            "aggregate": undefined,
            "className": "",
            "filterAll": false,
            "filterMethod": undefined,
            "filterable": undefined,
            "footerClassName": "",
            "footerStyle": Object {},
            "getFooterProps": [Function],
            "getHeaderProps": [Function],
            "getProps": [Function],
            "headerClassName": "",
            "headerStyle": Object {},
            "minWidth": 100,
            "resizable": undefined,
            "show": true,
            "sortMethod": undefined,
            "sortable": undefined,
            "style": Object {},
          },
          "columns": Array [
            Object {
              "Header": "column1",
              "accessor": "column1",
            },
            Object {
              "Header": "column2",
              "accessor": "column2",
            },
            Object {
              "Header": "column3",
              "accessor": "column3",
            },
          ],
          "data": Array [
            Object {
              "column1": 11,
              "column2": 12,
              "column3": "abc",
            },
            Object {
              "column1": 21,
              "column2": 22,
              "column3": "def",
            },
          ],
          "defaultExpanded": Object {},
          "defaultFilterMethod": [Function],
          "defaultFiltered": Array [],
          "defaultPageSize": 10,
          "defaultResized": Array [],
          "defaultSortDesc": false,
          "defaultSortMethod": [Function],
          "defaultSorted": Array [],
          "expanderDefaults": Object {
            "filterable": false,
            "resizable": false,
            "sortable": false,
            "width": 35,
          },
          "filterable": false,
          "freezeWhenExpanded": false,
          "getLoadingProps": [Function],
          "getNoDataProps": [Function],
          "getPaginationProps": [Function],
          "getProps": [Function],
          "getResizerProps": [Function],
          "getTableProps": [Function],
          "getTbodyProps": [Function],
          "getTdProps": [Function],
          "getTfootProps": [Function],
          "getTfootTdProps": [Function],
          "getTfootTrProps": [Function],
          "getTheadFilterProps": [Function],
          "getTheadFilterThProps": [Function],
          "getTheadFilterTrProps": [Function],
          "getTheadGroupProps": [Function],
          "getTheadGroupThProps": [Function],
          "getTheadGroupTrProps": [Function],
          "getTheadProps": [Function],
          "getTheadThProps": [Function],
          "getTheadTrProps": [Function],
          "getTrGroupProps": [Function],
          "getTrProps": [Function],
          "groupedByPivotKey": "_groupedByPivot",
          "indexKey": "_index",
          "loading": false,
          "loadingText": "Loading...",
          "multiSort": true,
          "nestingLevelKey": "_nestingLevel",
          "nextText": "Next",
          "noDataText": "No rows found",
          "ofText": "of",
          "onExpandedChange": undefined,
          "onFetchData": [Function],
          "onFilteredChange": undefined,
          "onPageChange": undefined,
          "onPageSizeChange": undefined,
          "onResizedChange": undefined,
          "onSortedChange": undefined,
          "originalKey": "_original",
          "pageSizeOptions": Array [
            5,
            10,
            20,
            25,
            50,
            100,
          ],
          "pageText": "Page",
          "pivotBy": undefined,
          "pivotDefaults": Object {},
          "pivotIDKey": "_pivotID",
          "pivotValKey": "_pivotVal",
          "previousText": "Previous",
          "resizable": true,
          "resolveData": [Function],
          "rowsText": "rows",
          "showPageJump": true,
          "showPageSizeOptions": true,
          "showPagination": true,
          "showPaginationBottom": false,
          "showPaginationTop": false,
          "sortable": true,
          "style": Object {},
          "subRowsKey": "_subRows",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h4
            style={
              Object {
                "textAlign": "center",
              }
            }
          >
            testable test table
          </h4>,
          <ReactTable
            AggregatedComponent={[Function]}
            ExpanderComponent={[Function]}
            FilterComponent={[Function]}
            LoadingComponent={[Function]}
            NextComponent={undefined}
            NoDataComponent={[Function]}
            PadRowComponent={[Function]}
            PaginationComponent={[Function]}
            PivotComponent={undefined}
            PivotValueComponent={[Function]}
            PreviousComponent={undefined}
            ResizerComponent={[Function]}
            TableComponent={[Function]}
            TbodyComponent={[Function]}
            TdComponent={[Function]}
            TfootComponent={[Function]}
            ThComponent={[Function]}
            TheadComponent={[Function]}
            TrComponent={[Function]}
            TrGroupComponent={[Function]}
            aggregatedKey="_aggregated"
            className="allow-wrap compact"
            collapseOnDataChange={true}
            collapseOnPageChange={true}
            collapseOnSortingChange={true}
            column={
              Object {
                "Aggregated": undefined,
                "Cell": undefined,
                "Expander": undefined,
                "Filter": undefined,
                "Footer": undefined,
                "Header": undefined,
                "Pivot": undefined,
                "PivotValue": undefined,
                "aggregate": undefined,
                "className": "",
                "filterAll": false,
                "filterMethod": undefined,
                "filterable": undefined,
                "footerClassName": "",
                "footerStyle": Object {},
                "getFooterProps": [Function],
                "getHeaderProps": [Function],
                "getProps": [Function],
                "headerClassName": "",
                "headerStyle": Object {},
                "minWidth": 100,
                "resizable": undefined,
                "show": true,
                "sortMethod": undefined,
                "sortable": undefined,
                "style": Object {},
              }
            }
            columns={
              Array [
                Object {
                  "Header": "column1",
                  "accessor": "column1",
                },
                Object {
                  "Header": "column2",
                  "accessor": "column2",
                },
                Object {
                  "Header": "column3",
                  "accessor": "column3",
                },
              ]
            }
            data={
              Array [
                Object {
                  "column1": 11,
                  "column2": 12,
                  "column3": "abc",
                },
                Object {
                  "column1": 21,
                  "column2": 22,
                  "column3": "def",
                },
              ]
            }
            defaultExpanded={Object {}}
            defaultFilterMethod={[Function]}
            defaultFiltered={Array []}
            defaultPageSize={10}
            defaultResized={Array []}
            defaultSortDesc={false}
            defaultSortMethod={[Function]}
            defaultSorted={Array []}
            expanderDefaults={
              Object {
                "filterable": false,
                "resizable": false,
                "sortable": false,
                "width": 35,
              }
            }
            filterable={false}
            freezeWhenExpanded={false}
            getLoadingProps={[Function]}
            getNoDataProps={[Function]}
            getPaginationProps={[Function]}
            getProps={[Function]}
            getResizerProps={[Function]}
            getTableProps={[Function]}
            getTbodyProps={[Function]}
            getTdProps={[Function]}
            getTfootProps={[Function]}
            getTfootTdProps={[Function]}
            getTfootTrProps={[Function]}
            getTheadFilterProps={[Function]}
            getTheadFilterThProps={[Function]}
            getTheadFilterTrProps={[Function]}
            getTheadGroupProps={[Function]}
            getTheadGroupThProps={[Function]}
            getTheadGroupTrProps={[Function]}
            getTheadProps={[Function]}
            getTheadThProps={[Function]}
            getTheadTrProps={[Function]}
            getTrGroupProps={[Function]}
            getTrProps={[Function]}
            groupedByPivotKey="_groupedByPivot"
            indexKey="_index"
            loading={false}
            loadingText="Loading..."
            multiSort={true}
            nestingLevelKey="_nestingLevel"
            nextText="Next"
            noDataText="No rows found"
            ofText="of"
            onExpandedChange={undefined}
            onFetchData={[Function]}
            onFilteredChange={undefined}
            onPageChange={undefined}
            onPageSizeChange={undefined}
            onResizedChange={undefined}
            onSortedChange={undefined}
            originalKey="_original"
            pageSizeOptions={
              Array [
                5,
                10,
                20,
                25,
                50,
                100,
              ]
            }
            pageText="Page"
            pivotBy={undefined}
            pivotDefaults={Object {}}
            pivotIDKey="_pivotID"
            pivotValKey="_pivotVal"
            previousText="Previous"
            resizable={true}
            resolveData={[Function]}
            rowsText="rows"
            showPageJump={true}
            showPageSizeOptions={true}
            showPagination={true}
            showPaginationBottom={false}
            showPaginationTop={false}
            sortable={true}
            style={Object {}}
            subRowsKey="_subRows"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "testable test table",
            "style": Object {
              "textAlign": "center",
            },
          },
          "ref": null,
          "rendered": "testable test table",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "AggregatedComponent": [Function],
            "ExpanderComponent": [Function],
            "FilterComponent": [Function],
            "LoadingComponent": [Function],
            "NextComponent": undefined,
            "NoDataComponent": [Function],
            "PadRowComponent": [Function],
            "PaginationComponent": [Function],
            "PivotComponent": undefined,
            "PivotValueComponent": [Function],
            "PreviousComponent": undefined,
            "ResizerComponent": [Function],
            "TableComponent": [Function],
            "TbodyComponent": [Function],
            "TdComponent": [Function],
            "TfootComponent": [Function],
            "ThComponent": [Function],
            "TheadComponent": [Function],
            "TrComponent": [Function],
            "TrGroupComponent": [Function],
            "aggregatedKey": "_aggregated",
            "className": "allow-wrap compact",
            "collapseOnDataChange": true,
            "collapseOnPageChange": true,
            "collapseOnSortingChange": true,
            "column": Object {
              "Aggregated": undefined,
              "Cell": undefined,
              "Expander": undefined,
              "Filter": undefined,
              "Footer": undefined,
              "Header": undefined,
              "Pivot": undefined,
              "PivotValue": undefined,
              "aggregate": undefined,
              "className": "",
              "filterAll": false,
              "filterMethod": undefined,
              "filterable": undefined,
              "footerClassName": "",
              "footerStyle": Object {},
              "getFooterProps": [Function],
              "getHeaderProps": [Function],
              "getProps": [Function],
              "headerClassName": "",
              "headerStyle": Object {},
              "minWidth": 100,
              "resizable": undefined,
              "show": true,
              "sortMethod": undefined,
              "sortable": undefined,
              "style": Object {},
            },
            "columns": Array [
              Object {
                "Header": "column1",
                "accessor": "column1",
              },
              Object {
                "Header": "column2",
                "accessor": "column2",
              },
              Object {
                "Header": "column3",
                "accessor": "column3",
              },
            ],
            "data": Array [
              Object {
                "column1": 11,
                "column2": 12,
                "column3": "abc",
              },
              Object {
                "column1": 21,
                "column2": 22,
                "column3": "def",
              },
            ],
            "defaultExpanded": Object {},
            "defaultFilterMethod": [Function],
            "defaultFiltered": Array [],
            "defaultPageSize": 10,
            "defaultResized": Array [],
            "defaultSortDesc": false,
            "defaultSortMethod": [Function],
            "defaultSorted": Array [],
            "expanderDefaults": Object {
              "filterable": false,
              "resizable": false,
              "sortable": false,
              "width": 35,
            },
            "filterable": false,
            "freezeWhenExpanded": false,
            "getLoadingProps": [Function],
            "getNoDataProps": [Function],
            "getPaginationProps": [Function],
            "getProps": [Function],
            "getResizerProps": [Function],
            "getTableProps": [Function],
            "getTbodyProps": [Function],
            "getTdProps": [Function],
            "getTfootProps": [Function],
            "getTfootTdProps": [Function],
            "getTfootTrProps": [Function],
            "getTheadFilterProps": [Function],
            "getTheadFilterThProps": [Function],
            "getTheadFilterTrProps": [Function],
            "getTheadGroupProps": [Function],
            "getTheadGroupThProps": [Function],
            "getTheadGroupTrProps": [Function],
            "getTheadProps": [Function],
            "getTheadThProps": [Function],
            "getTheadTrProps": [Function],
            "getTrGroupProps": [Function],
            "getTrProps": [Function],
            "groupedByPivotKey": "_groupedByPivot",
            "indexKey": "_index",
            "loading": false,
            "loadingText": "Loading...",
            "multiSort": true,
            "nestingLevelKey": "_nestingLevel",
            "nextText": "Next",
            "noDataText": "No rows found",
            "ofText": "of",
            "onExpandedChange": undefined,
            "onFetchData": [Function],
            "onFilteredChange": undefined,
            "onPageChange": undefined,
            "onPageSizeChange": undefined,
            "onResizedChange": undefined,
            "onSortedChange": undefined,
            "originalKey": "_original",
            "pageSizeOptions": Array [
              5,
              10,
              20,
              25,
              50,
              100,
            ],
            "pageText": "Page",
            "pivotBy": undefined,
            "pivotDefaults": Object {},
            "pivotIDKey": "_pivotID",
            "pivotValKey": "_pivotVal",
            "previousText": "Previous",
            "resizable": true,
            "resolveData": [Function],
            "rowsText": "rows",
            "showPageJump": true,
            "showPageSizeOptions": true,
            "showPagination": true,
            "showPaginationBottom": false,
            "showPaginationTop": false,
            "sortable": true,
            "style": Object {},
            "subRowsKey": "_subRows",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
